docs/enhancements/cluster-pool-inventory.md

```diff
@@ -0,0 +1,158 @@
+Like maybe for baremetal, we need to supply a preconfigured hardware inventory for each cluster, but it's okay to generate the names.
+
+### `Size` and `MaxSize`
+If `Inventory` is used, `ClusterPool.Spec.Size` and `.MaxSize` are implicitly constrained to the length of the list of nonempty `ClusterDeployments[].Names`.
+Setting either/both to a smaller number still works as you would expect.
+
+### Pool Version
+To make [adding](#adding-an-inventory) and [deleting](#deleting-the-inventory) `Inventory` work sanely, we will adjust the computation of the pool version used for [stale CD detection/replacement](https://issues.redhat.com/browse/HIVE-1058) as follows:
+- When `Inventory` is present, append an arbitrary fixed string before the [final hash operation](https://github.com/openshift/hive/blob/0b9229b91e6f8a3c2bf095efbdf017226e69d026/pkg/controller/clusterpool/clusterpool_controller.go#L479).
+  (We don't want to recalculate the hash any time the inventory changes, as this would treat *all* existing CDs as stale and replace them.)
+- When `Inventory` is absent, compute the pool version as before.
+  This ensures the version does not change (triggering replacement of all unclaimed CDs) for existing pools when hive is upgraded to include this feature.
+
+### Handling Inventory Updates
+#### Adding An Inventory
+Adding an `Inventory` to a ClusterPool which previously didn't have one will cause the controller to [recompute the pool version](#pool-version), rendering all existing unclaimed clusters stale, causing them to be replaced gradually.
+
+#### Adding An Entry
+If `MaxSize` is unset or exceeds the length of the inventory, and `Size` allows, adding a new entry will cause a new ClusterDeployment to be added to the pool.
+If the pool is already at `[Max]Size` there is no immediate effect.
+
+#### Removing An Entry
+- If the entry is unused (no pool CD with that name exists), this is a no-op.
+- If an _unclaimed_ CD exists with that name, we delete it.
+  The controller will replace it, assuming an unused entry is available.
+- If a _claimed_ CD exists with that name, no-op.
+
+These are conceived to correlate as closely as possible to what happens when editing a pool's `Size`.
+
+#### Deleting The Inventory
+This will change the [pool version](#pool-version), rendering existing unclaimed clusters stale and causing the controller to replace them gradually.
+Absent the inventory, the replacements will have generated names as "normal" (before this feature).
+The administrator may wish to speed up this process by manually deleting CDs, or scaling the pool size to zero and back.
+
+### Challenge: Namespace Collision
+Today, ClusterPool uses the same Name and Namespace for a given ClusterDeployment.
+There is code that relies on these being the same.
+If we continue this practice, a user-specified inventory has the potential to contain a name that would conflict with an existing namespace.
```

Comment:
I think you have some flexibility in terms of what the clusterdeployment is named, and the actual cluster name fields in the clusterdeployment. (spec.clusterName, spec.baseName) Could you work out a way to take the given name and append something random when naming your cd and namespace, even though the cluster name will be consistent and what the user requests in their inventory?

Reply:
Yes!

The simplest thing, without mucking with [Builder.Name](https://github.com/openshift/hive/blob/8671ed2067e47e5ed0b6e718e7163a6741ed7b39/pkg/clusterresource/builder.go#L27-L30), would be to keep [using Builder as is](https://github.com/openshift/hive/blob/8671ed2067e47e5ed0b6e718e7163a6741ed7b39/pkg/controller/clusterpool/clusterpool_controller.go#L665-L679) and then change the cd.spec.clusterName at the same time we [set the pool reference](https://github.com/openshift/hive/blob/8671ed2067e47e5ed0b6e718e7163a6741ed7b39/pkg/controller/clusterpool/clusterpool_controller.go#L700-L701).

This solves the namespace collision problem nicely... but doesn't help with the cluster name concurrency problem discussed [above](https://github.com/openshift/hive/pull/1549#discussion_r716593307).